STYLE_GUIDE.md

```diff
@@ -155,30 +142,67 @@ if parameter is None:
 ```
 
 
-## 5. Imports and `__init__.py`
+## 6. Imports and `__init__.py`
 
 All imports in composer should be absolute -- that is, they do not begin with a period.
-In addition, all imports should be added into an appropriate section of [setup.py](setup.py).
 
-### 5.1 Optional Dependencies
+### 6.1 External Dependencies
+1.  All external dependencies must be specified in both [setup.py](setup.py) for pip and [meta.yaml](meta.yaml)
+    for Anaconda.
 
-If an import is not in `install_requires`, then this import MUST be conditionally imported in the code
--- e.g. like this:
+1.  If a dependency is not core to Composer (e.g. it is for a model, dataset, algorithm, or some callbacks):
+    1.  It must be specified in a entry of the `extra_deps` dictionary of [setup.py](setup.py).
+        This dictionary groups dependencies that can be conditionally installed. An entry named `foo`
+        can be installed with `pip install mosaicml[foo]`. For example, running `pip install mosaicml[unet]`
+        will install everything in `install_requires`, along with `monai` and `scikit-learn`.
+    1.  It must also be specified in the `run_constrained` and the `test.requires` section.
+    1.  The import must be conditionally imported in the code. For example:
+
+        ```python
+        def unet():
+            try:
+                import monai
+            except ImportError as e:
+                raise ImportError(textwrap.dedent("""\
+                    Composer was installed without unet support. To use unet with Composer, run: `pip install mosaicml
+                    [unet]` if using pip or `conda install -c conda-forge monai` if using Anaconda""") from e
+        ```
+
+        This style allows users to perform minimal install of Composer without triggering `ImportError`s if
+        an optional dependency is missing.
+
+    1.  If the dependency is core to Composer, add the dependency to the `install_requires` section of
+        [setup.py](./setup.py) and the `requirements.run` section of [meta.yaml](./meta.yaml).
+
+### 6.2 Use of `__all__`
+
+All public modules must define `__all__` to be the list of members that should be re-exported.
+The variable is necessary to limit what `from XXX import *` imports, and so the documentation build
+only documents exported members (and not unrelated re-imports).
```

Comment:
```suggestion
The variable is necessary to 1) limit what `from XXX import *` imports, and 2) ensure that the documentation only includes exported members, not unrelated re-imports.
```
A little hard to parse. Expected "and so" to elaborate on for first clause, not introduce a second list item; and expected "build" to be a verb initially.